lib/internal/modules/esm/resolve.js

```diff
@@ -1,26 +1,604 @@
 'use strict';
 
 const {
+  ArrayIsArray,
+  JSONParse,
+  JSONStringify,
+  ObjectIs,
+  ObjectGetOwnPropertyNames,
+  ObjectPrototypeHasOwnProperty,
   SafeMap,
+  StringPrototypeEndsWith,
+  StringPrototypeIncludes,
+  StringPrototypeIndexOf,
+  StringPrototypeStartsWith,
+  StringPrototypeSubstr,
 } = primordials;
 
 const internalFS = require('internal/fs/utils');
 const { NativeModule } = require('internal/bootstrap/loaders');
-const { realpathSync } = require('fs');
+const {
+  closeSync,
+  fstatSync,
+  openSync,
+  readFileSync,
+  realpathSync,
+  statSync,
+  Stats,
+} = require('fs');
 const { getOptionValue } = require('internal/options');
 const { sep } = require('path');
 
 const preserveSymlinks = getOptionValue('--preserve-symlinks');
 const preserveSymlinksMain = getOptionValue('--preserve-symlinks-main');
 const typeFlag = getOptionValue('--input-type');
-const { resolve: moduleWrapResolve } = internalBinding('module_wrap');
 const { URL, pathToFileURL, fileURLToPath } = require('internal/url');
-const { ERR_INPUT_TYPE_NOT_ALLOWED,
-        ERR_UNSUPPORTED_ESM_URL_SCHEME } = require('internal/errors').codes;
+const {
+  ERR_INPUT_TYPE_NOT_ALLOWED,
+  ERR_INVALID_MODULE_SPECIFIER,
+  ERR_INVALID_PACKAGE_CONFIG,
+  ERR_INVALID_PACKAGE_TARGET,
+  ERR_MODULE_NOT_FOUND,
+  ERR_PACKAGE_PATH_NOT_EXPORTED,
+  ERR_UNSUPPORTED_ESM_URL_SCHEME,
+} = require('internal/errors').codes;
 
 const realpathCache = new SafeMap();
+const packageJSONCache = new SafeMap();  /* string -> PackageConfig */
+
+function tryStatSync(path) {
+  try {
+    return statSync(path);
+  } catch {
+    return new Stats();
+  }
+}
+
+function readIfFile(path) {
+  let fd;
+  try {
+    fd = openSync(path, 'r');
+  } catch {
+    return undefined;
+  }
+  try {
+    if (!fstatSync(fd).isFile()) return undefined;
```

Comment:
We should make this cached, but I want that to be a different PR since we don't have a good cache system for builtin prototype methods currently.